clash_lib/src/proxy/utils/test_utils/docker_runner.rs

```diff
@@ -0,0 +1,158 @@
+            .await?;
+
+        let id = docker
+            .create_container::<&str, Z>(None, container_conf)
+            .await?
+            .id;
+        docker.start_container::<String>(&id, None).await?;
+        Ok(Self {
+            _instance: docker,
+            id,
+        })
+    }
+
+    #[allow(unused)]
+    pub async fn exec(&self, cmd: Vec<&str>) -> anyhow::Result<()> {
+        // non interactive
+        let exec = self
+            ._instance
+            .create_exec(
+                &self.id,
+                CreateExecOptions {
+                    attach_stdout: Some(true),
+                    attach_stderr: Some(true),
+                    cmd: Some(cmd),
+                    ..Default::default()
+                },
+            )
+            .await?
+            .id;
+        if let StartExecResults::Attached { mut output, .. } =
+            self._instance.start_exec(&exec, None).await?
+        {
+            while let Some(Ok(msg)) = output.next().await {
+                print!("{msg}");
+            }
+            return Ok(());
+        } else {
+            anyhow::bail!("failed to execute cmd")
+        }
+    }
+
+    // will make sure the container is cleaned up after the future is finished
+    pub async fn run_and_cleanup(
+        self,
+        f: impl Future<Output = anyhow::Result<()>> 
+    ) -> anyhow::Result<()> {
+        let fut = Box::pin(f);
+        let res = fut.await;
+        // make sure the container is cleaned up
+        // TODO: select a timeout future as well, make sure it can quit smoothly
+        self.cleanup().await?;
+
+        res
+    }
+
+    // you can run the cleanup manually
+    pub async fn cleanup(self) -> anyhow::Result<()> {
+        let s = self
+            ._instance
+            .remove_container(
+                &self.id,
+                Some(RemoveContainerOptions {
+                    force: true,
+                    ..Default::default()
+                }),
+            )
+            .await?;
+        Ok(s)
+    }
+}
+
+pub fn mount_config(pairs: &[(&str, &str)]) -> Vec<Mount> {
+    pairs
+        .iter()
+        .map(|(src, dst)| Mount {
+            target: Some(dst.to_string()),
+            source: Some(src.to_string()),
+            typ: Some(bollard::secret::MountTypeEnum::BIND),
+            read_only: Some(false),
+            ..Default::default()
+        })
+        .collect::<Vec<_>>()
+}
+
+pub fn default_host_config() -> HostConfig {
+    let mut host_config = HostConfig::default();
+    // we need to use the host mode to enable the benchmark function
+    #[cfg(not(target_os = "macos"))]
+    {
+        host_config.network_mode = Some("host".to_owned());
+    }
+    host_config.port_bindings = Some(
+        [
+            (
+                "10002/tcp".to_owned(),
+                Some(vec![PortBinding {
+                    host_ip: Some("0.0.0.0".to_owned()),
+                    host_port: Some("10002".to_owned()),
```

Comment:
would there be port binding conflict when the tests were run in parallel? https://doc.rust-lang.org/book/ch11-02-running-tests.html#running-tests-in-parallel-or-consecutively

Reply:
Yep, I found that problem, too, after the push of this PR, since it utilize `#[tokio::test]`.

Will a atomic counter be helpful? So we can calculate the port at the beginning of each test. An there shall be no conflict.

Reply:
That should work too. 

Or we can maintain a hardcoded port lists for each test somewhere in the code and assign them while testing different protocols. As long as they are different. 

Which ever works and is simple to implement

Reply:
I prefer an auto-increment  & global atomic counter(as the port). It's easy to implement, too.

Reply:
I use serial_test instead. To change the port in different test is a bad idea, since the proxy server's port is hardcoded in the config file.

Reply:
yeah we can pre process the config files too with dyn ports, but that adds extra work. serial should be fien for now until the entire tests runs too long